control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs

```diff
@@ -181,3 +182,30 @@ async fn prune_volume_snapshot_reconciler(
 
     Ok(PollerState::Idle)
 }
+
+#[tracing::instrument(skip(snapshot, context), level = "trace", fields(snapshot.id = %snapshot.uuid(), request.reconcile = true))]
+async fn delete_no_transaction_volume_snapshot_reconciler(
+    snapshot: &mut OperationGuardArc<VolumeSnapshot>,
+    context: &PollContext,
+) -> PollResult {
+    // Only proceed if the spec is in creating and has no transactions.
+    if !snapshot.as_ref().status().creating()
+        && !snapshot.as_ref().metadata().transactions().is_empty()
+    {
+        return Ok(PollerState::Idle);
+    }
+
+    let Some(snap_rsc) = context.specs().volume_snapshot_rsc(snapshot.uuid()) else {
+        return Ok(PollerState::Idle);
+    };
```

Comment:
why do we need this, can't we just do:
```suggestion
    let snap_rsc = snapshot.deref().clone();
```

Reply:
Right. We can just do this.

Reply:
Then let's add a way to get it, and fix existing usages too, example:

```patch
diff --git a/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs b/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs
index 28c3a0c7..4868f649 100644
--- a/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs
+++ b/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs
@@ -4,7 +4,7 @@ use crate::{
         resources::{
             operations::{ResourceLifecycleWithLifetime, ResourcePruning},
             operations_helper::{OperationSequenceGuard, SpecOperationsHelper},
-            OperationGuardArc,
+            OperationGuardArc, ResourceMutex,
         },
         task_poller::{PollContext, PollResult, PollTimer, PollerState, TaskPoller},
     },
@@ -90,10 +90,7 @@ async fn deleting_volume_snapshot_reconciler(
         return Ok(PollerState::Idle);
     }
 
-    let Some(snap_rsc) = context.specs().volume_snapshot_rsc(snapshot.uuid()) else {
-        return Ok(PollerState::Idle);
-    };
-
+    let snap_rsc: ResourceMutex<VolumeSnapshot> = snapshot.resource().clone();
     let snap_uuid = snapshot.uuid().clone();
     match snapshot
         .destroy(
diff --git a/control-plane/agents/src/bin/core/controller/resources/mod.rs b/control-plane/agents/src/bin/core/controller/resources/mod.rs
index 3540e960..a1a3b41c 100644
--- a/control-plane/agents/src/bin/core/controller/resources/mod.rs
+++ b/control-plane/agents/src/bin/core/controller/resources/mod.rs
@@ -197,6 +197,11 @@ impl<T: OperationSequencer + Sized, R> OperationGuard<T, R> {
         }
     }
 }
+impl<T: Clone + Debug + AsOperationSequencer, R> OperationGuard<ResourceMutex<T>, R> {
+    pub fn resource(&self) -> &ResourceMutex<T> {
+        &self.inner
+    }
+}
 
 /// Tracing simple string messages with resource specific information
 /// eg, volume.uuid for volumes and replica.uuid for replicas
```

Reply:
Done.